pkg/apis/batch/validation/validation.go

```diff
@@ -400,6 +405,32 @@ func validateJobStatus(status *batch.JobStatus, fldPath *field.Path) field.Error
 	if status.Terminating != nil {
 		allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(*status.Terminating), fldPath.Child("terminating"))...)
 	}
+
+	// A strengthened validation of the Job status is needed since the
+	// Job managed-by label let's the Job object be controlled by external
+	// controllers. We want to make sure the status set by the external
+	// controllers meet the expectations of the clients of the Job API.
+	// In particular, we don't allow to set both Complete and Failed conditions.
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobManagedByLabel) {
+		isComplete := isConditionTrue(status.Conditions, batch.JobComplete)
+		if isComplete && isConditionTrue(status.Conditions, batch.JobFailed) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions"), field.OmitValueType{}, "both Complete=True and Failed=True conditions are set"))
+		}
+		if status.CompletionTime != nil && !isComplete {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("completionTime"), status.CompletionTime, "completionTime can only be set when Complete=True"))
+		}
+		if ptr.Deref(job.Spec.CompletionMode, batch.NonIndexedCompletion) == batch.NonIndexedCompletion {
+			if status.CompletedIndexes != "" {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("completedIndexes"), status.CompletedIndexes, "non-empty completedIndexes when non-indexed completion mode"))
+			}
+		}
+		if job.Spec.BackoffLimitPerIndex == nil {
```

Comment:
I assumed if the `BackoffLimitPerIndex` gate is disabled, we don't want people to use `status.FailedIndexes` I assume. If it is disabled, but you have `job.Spec.BackoffLimitPerIndex != nil` it is ok, because it means the feature gate was enabled at some point and disabled later.

Reply:
It makes sense. Thank you for the clarifications.